File: Algorithm/213. House Robber II.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-06-23 22:00:39
# @Last modified by:   WuLC
# @Last Modified time: 2016-06-23 22:00:53
# @Email: liangchaowu5@gmail.com

class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        dp1,dp2 = nums[:], nums[1:]
        dp1.insert(0, 0)
        dp2.insert(0, 0)
        for i in xrange(3, len(dp1)):
            if i != len(dp1)-1:
                dp2[i] = max(dp2[i]+dp2[i-2], dp2[i]+dp2[i-3])
            dp1[i] = max(dp1[i]+dp1[i-2], dp1[i]+dp1[i-3])
        if dp2[-1] > 0:
            dp1[-1] = dp2[-1]
        max_sum = dp1[0]
        for i in xrange(1, len(dp1)):
            if dp1[i] > max_sum:
                max_sum = dp1[i]
        return max_sum

File: Algorithm/Java/213. House Robber II.java
/*
* @Author: WuLC
* @Date:   2017-06-29 12:58:24
* @Last Modified by:   WuLC
* @Last Modified time: 2017-06-29 12:59:28
* @Email: liangchaowu5@gmail.com
*/


// two choices, choose to exclude the first house or th last house
public class Solution 
{
    public int rob(int[] nums) 
    {
        if(nums.length == 0) return 0;
        else if(nums.length == 1) return nums[0];
        else if(nums.length == 2) return Math.max(nums[0], nums[1]);
        else return Math.max(helper(nums, 0, nums.length -2), helper(nums, 1, nums.length - 1));
    }
    
    public int helper(int[] nums, int start, int end)
    {
        int pre1 = nums[start], pre2 = Math.max(nums[start], nums[start+1]);
        int index = start + 2, curr = pre2;
        while(index <= end)
        {
            curr = Math.max(pre2, pre1 + nums[index]);
            pre1 = pre2;
            pre2 = curr;
            index++;
        }
        return curr;
    }
}